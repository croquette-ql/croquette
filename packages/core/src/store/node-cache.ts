import type { DocumentNode, OperationDefinitionNode, SelectionSetNode, FragmentDefinitionNode } from 'graphql';
import { OperationIdGenerator } from '../types';
import { DefaultOperationIdGenerator } from './operation-id';

interface NodeReference {
  __ref: true;
  type: string;
  id: string;
}

type NormalizedDataAtom = NodeReference | string | number | boolean | null | undefined;
type NormalizedDataValue = NormalizedDataAtom | ReadonlyArray<NormalizedDataAtom>;
type NormalizedDataRecord = { __typename: string } & Record<string, NormalizedDataValue>;
type NormalizedData = Record<string, NormalizedDataRecord>;

export type NodeCacheCreateOptions = {
  initialData?: {
    operationResults: Record<string, string>;
    normalizedData: NormalizedData;
  };
  idGenerator?: OperationIdGenerator;
};

export type CacheResult = {
  data: any;
  missingDataLocations: string[];
};

function isNodeReference(x: any): x is NodeReference {
  return typeof x === 'object' && x.__ref === true;
}

function findQueryOperationASTNode(ast: DocumentNode) {
  const hit = ast.definitions.find(node => {
    return node.kind === 'OperationDefinition' && node.operation === 'query';
  });
  if (!hit) {
    throw new Error('no query');
  }
  return hit as OperationDefinitionNode;
}

function getFragmentDefinitionMap(ast: DocumentNode) {
  const map = new Map<string, FragmentDefinitionNode>();
  ast.definitions.forEach(def => {
    if (def.kind === 'FragmentDefinition') {
      map.set(def.name.value, def);
    }
  });
  return map;
}

export class NodeCache {
  private _normalizedData: NormalizedData;
  private _operationResults: Record<string, string>;
  private _operationIdGenerator: OperationIdGenerator;

  constructor(options: NodeCacheCreateOptions = {}) {
    this._normalizedData = options.initialData?.normalizedData ?? {};
    this._operationResults = options.initialData?.operationResults ?? {};
    this._operationIdGenerator = options.idGenerator ?? new DefaultOperationIdGenerator();
  }

  serialize() {
    return {
      normalizedData: this._normalizedData,
      operationResults: this._operationResults,
    };
  }

  readQuery({ query, variables }: { query: DocumentNode; variables: any }) {
    const operationId = this._operationIdGenerator.getOperationId({ query, variables });
    const operationResultKey = this._operationResults[operationId];
    if (!operationResultKey) {
      return {
        data: null,
        missingDataLocations: [], // FIXME
      };
    }
    const queryAstNode = findQueryOperationASTNode(query);
    const fragmentMap = getFragmentDefinitionMap(query);
    const missingDataLocations: string[] = [];

    const record = this._normalizedData[operationResultKey];
    const selectionSet = queryAstNode.selectionSet;
    const getSelectionData = (selectionSet: SelectionSetNode, targetRecord: any) => {
      let ret: any = {};
      selectionSet.selections.forEach(selection => {
        if (selection.kind === 'Field') {
          const value = targetRecord[selection.name.value];
          if (selection.selectionSet) {
            if (isNodeReference(value)) {
              const nextRecord = this._normalizedData[`${value.type}:${value.id}`];
              ret[selection.name.value] = getSelectionData(selection.selectionSet, nextRecord);
            } else if (Array.isArray(value)) {
              ret[selection.name.value] = value.map(v => {
                if (isNodeReference(v)) {
                  const nextRecord = this._normalizedData[`${v.type}:${v.id}`];
                  return getSelectionData(selection.selectionSet!, nextRecord);
                } else {
                  throw new Error('Illegal selection');
                }
              });
            } else {
              throw new Error('Illegal selection');
            }
          } else if (typeof value !== 'undefined') {
            ret[selection.name.value] = value;
          } else {
            missingDataLocations.push(selection.name.value);
          }
        } else if (selection.kind === 'FragmentSpread') {
          const fragmentDef = fragmentMap.get(selection.name.value);
          if (!fragmentDef) {
            throw new Error(`cannot find fragment definition for ${selection.name.value}`);
          }
          ret = { ...ret, ...getSelectionData(fragmentDef.selectionSet, targetRecord) };
        }
      });
      return ret;
    };
    const data = getSelectionData(selectionSet, record);

    if (missingDataLocations.length > 0) {
      return {
        data: null,
        missingDataLocations,
      };
    } else {
      return {
        data,
        missingDataLocations: [],
      };
    }
  }

  writeQuery({ query, variables }: { query: DocumentNode; variables: any }, data: any) {
    const operationId = this._operationIdGenerator.getOperationId({ query, variables });
    const queryAstNode = findQueryOperationASTNode(query);
    const fragmentMap = getFragmentDefinitionMap(query);
    const writeSelectionData = (selectionSet: SelectionSetNode, targetObject: any, location: (string | number)[]) => {
      const typename = targetObject.__typename;
      if (!typename) {
        return null;
      }
      const objectId = targetObject.id ?? [operationId, ...location].join('/');
      const id = `${typename}:${objectId}`;
      let base: NormalizedDataRecord;
      if (this._normalizedData[id]) {
        base = this._normalizedData[id];
      } else {
        base = {
          __typename: typename,
        };
        this._normalizedData[id] = base;
      }
      selectionSet.selections.forEach(selection => {
        if (selection.kind === 'Field') {
          const fieldName = selection.name.value;
          const value = targetObject[fieldName];
          if (selection.selectionSet) {
            if (Array.isArray(value)) {
              base[fieldName] = value.map((v, i) => {
                return writeSelectionData(selection.selectionSet!, v, [...location, fieldName, i]);
              });
            } else if (value != null) {
              base[fieldName] = writeSelectionData(selection.selectionSet, value, [...location, fieldName]);
            }
          } else {
            base[fieldName] = value;
          }
        } else if (selection.kind === 'FragmentSpread') {
          const fragmentDef = fragmentMap.get(selection.name.value);
          if (!fragmentDef) {
            throw new Error(`cannot find fragment definition for ${selection.name.value}`);
          }
          writeSelectionData(fragmentDef.selectionSet, targetObject, location);
        }
      });
      const ref: NodeReference = {
        __ref: true,
        id: objectId,
        type: typename,
      };
      return ref;
    };
    const selectionSet = queryAstNode.selectionSet;
    const queryRef = writeSelectionData(selectionSet, data, []);
    if (queryRef) {
      this._operationResults[operationId] = `${queryRef.type}:${queryRef.id}`;
    }
  }
}
