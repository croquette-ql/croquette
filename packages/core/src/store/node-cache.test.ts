import { parse } from 'graphql';
import { NodeCache } from './node-cache';

describe(NodeCache, () => {
  describe('selection pattern', () => {
    describe(NodeCache.prototype.readQuery, () => {
      it('should return data when query has flat scalar selection', () => {
        const cache = new NodeCache({
          initialData: {
            operationResults: {
              'MyQuery/{}': 'Query:0',
            },
            normalizedData: {
              'Query:0': {
                __typename: 'Query',
                stringValue: 'hello',
                intValue: 1,
                floatValue: 1.0,
                boolValue: false,
                nullableValue: null,
              },
            },
          },
        });
        const query = `
          query MyQuery {
            stringValue
            intValue
            floatValue
            boolValue
            nullableValue
          }
        `;
        expect(cache.readQuery({ query: parse(query), variables: {} })).toStrictEqual({
          data: {
            stringValue: 'hello',
            intValue: 1,
            floatValue: 1.0,
            boolValue: false,
            nullableValue: null,
          },
          missingDataLocations: [],
        });
      });

      it('should return data when query has nested object selection', () => {
        const cache = new NodeCache({
          initialData: {
            operationResults: {
              'MyQuery/{}': 'Query:0',
            },
            normalizedData: {
              'Query:0': {
                __typename: 'Query',
                greeting: { __ref: true, type: 'Greeting', id: '0' },
              },
              'Greeting:0': {
                __typename: 'Greeting',
                hello: 'world',
              },
            },
          },
        });
        const query = `
          query MyQuery {
            greeting {
              hello
            }
          }
        `;
        expect(cache.readQuery({ query: parse(query), variables: {} })).toStrictEqual({
          data: {
            greeting: {
              hello: 'world',
            },
          },
          missingDataLocations: [],
        });
      });

      it('should return data when query has array selection', () => {
        const cache = new NodeCache({
          initialData: {
            operationResults: {
              'MyQuery/{}': 'Query:0',
            },
            normalizedData: {
              'Query:0': {
                __typename: 'Query',
                nodes: [
                  { __ref: true, type: 'Node', id: '0' },
                  { __ref: true, type: 'Node', id: '1' },
                ],
              },
              'Node:0': {
                __typename: 'Node',
                name: 'hoge',
              },
              'Node:1': {
                __typename: 'Node',
                name: 'fuga',
              },
            },
          },
        });
        const query = `
          query MyQuery {
            nodes {
              name
            }
          }
        `;
        expect(cache.readQuery({ query: parse(query), variables: {} })).toStrictEqual({
          data: {
            nodes: [{ name: 'hoge' }, { name: 'fuga' }],
          },
          missingDataLocations: [],
        });
      });

      it('should return data when query has fargment spread', () => {
        const cache = new NodeCache({
          initialData: {
            operationResults: {
              'MyQuery/{}': 'Query:0',
            },
            normalizedData: {
              'Query:0': {
                __typename: 'Query',
                hello: 'world',
              },
            },
          },
        });
        const query = `
          fragment MyFragment on Query {
            hello
          }
          query MyQuery {
            ...MyFragment
          }
        `;
        expect(cache.readQuery({ query: parse(query), variables: {} })).toStrictEqual({
          data: {
            hello: 'world',
          },
          missingDataLocations: [],
        });
      });
    });
  });
});
