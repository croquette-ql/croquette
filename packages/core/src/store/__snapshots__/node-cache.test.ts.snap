// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeCache selection pattern writeQuery should write data when query has array selection 1`] = `
Object {
  "normalizedData": Object {
    "Node:MyQuery/{}/nodes/0": Object {
      "__typename": "Node",
      "name": "hoge",
    },
    "Node:MyQuery/{}/nodes/1": Object {
      "__typename": "Node",
      "name": "fuga",
    },
    "Query:MyQuery/{}": Object {
      "__typename": "Query",
      "nodes": Array [
        Object {
          "__ref": true,
          "id": "MyQuery/{}/nodes/0",
          "type": "Node",
        },
        Object {
          "__ref": true,
          "id": "MyQuery/{}/nodes/1",
          "type": "Node",
        },
      ],
    },
  },
  "operationResults": Object {
    "MyQuery/{}": "Query:MyQuery/{}",
  },
}
`;

exports[`NodeCache selection pattern writeQuery should write data when query has fragment spread 1`] = `
Object {
  "normalizedData": Object {
    "Query:MyQuery/{}": Object {
      "__typename": "Query",
      "hello": "world",
    },
  },
  "operationResults": Object {
    "MyQuery/{}": "Query:MyQuery/{}",
  },
}
`;

exports[`NodeCache selection pattern writeQuery should write data when query has nested object selection 1`] = `
Object {
  "normalizedData": Object {
    "Greeting:MyQuery/{}/greeting": Object {
      "__typename": "Greeting",
      "hello": "world",
    },
    "Query:MyQuery/{}": Object {
      "__typename": "Query",
      "greeting": Object {
        "__ref": true,
        "id": "MyQuery/{}/greeting",
        "type": "Greeting",
      },
    },
  },
  "operationResults": Object {
    "MyQuery/{}": "Query:MyQuery/{}",
  },
}
`;

exports[`NodeCache selection pattern writeQuery should write data when query has nested object selection with id field 1`] = `
Object {
  "normalizedData": Object {
    "Greeting:0": Object {
      "__typename": "Greeting",
      "hello": "world",
      "id": "0",
    },
    "Query:MyQuery/{}": Object {
      "__typename": "Query",
      "greeting": Object {
        "__ref": true,
        "id": "0",
        "type": "Greeting",
      },
    },
  },
  "operationResults": Object {
    "MyQuery/{}": "Query:MyQuery/{}",
  },
}
`;

exports[`NodeCache selection pattern writeQuery should write data with query 1`] = `
Object {
  "normalizedData": Object {
    "Query:MyQuery/{}": Object {
      "__typename": "Query",
      "boolValue": false,
      "floatValue": 1,
      "intValue": 1,
      "nullableValue": null,
      "stringValue": "hello",
    },
  },
  "operationResults": Object {
    "MyQuery/{}": "Query:MyQuery/{}",
  },
}
`;
